docs/mkdocs/docs/features/namespace.md

```diff
@@ -0,0 +1,52 @@
+# `nlohmann` Namespace
+
+The 3.11.0 release introduced an inline namespace to allow different parts of a codebase to use different versions of
+the library as long as they never exchange instances of library types.
+
+In the 3.11.2 release, that namespace was split into the tag namespace – encoding the state of ABI-affecting macros –
+and the version namespace – encoding the version of the library.
+
+The full default namespace name is derived as follows:
+
+- The root namespace is always `nlohmann`.
+- The tag namespace starts with `nlohmann_json` and is followed by serveral optional ABI tags according to the value of
```

Comment:
Just thinking out loud here.  I wonder if something like `features` (or `abi_tags`) instead of `nlohmann_json` would make it clearer what is being enumerated here.  `nlohmann::features_diag_ldvcmp::v3_11_2` for example.  Might look a bit weird with no features though: `nlohmann::features::v3_11_2`.  Would it make sense to exclude this namespace if no features were enabled?  That combined with excluding the version namespace would result in just the `nlohmann` namespace as we had before, might not be what we want.

Reply:
I don't have an opinion about whether I like `features`, `abi_tags` or anything else more - but I do like the idea of "collapsing" everything back to `nlohmann::` when ABI tags and versions are switched off. This way, the new approach feels like an organic addition rather than a drastic change than needs much explanation.

(There, I think `abi_tags` does feel like the better prefix...)

Reply:
Originally, I wanted `json` to be part of the name, to make it clear what library the types belong to. This is ultimately not important, as the types still appear in the top-level namespace.

While it would be quite elegant to collapse the inline namespaces if the version is disabled and no ABI-altering options are enabled, it would also eliminate the whole point of the tag namespace. A 3.11.x+ version of the library with the aforementioned configuration is potentially different from an old version with an unknown configuration (because it is not part of the symbol names). To prevent linking errors in this scenario, the tag namespace should still be part of the symbol names.

What I can be convinced of is to undo the split of the namespace, which would result in the following hybrid:
```
nlohmann::nlohmann_json[_diag][_ldvcmp][_v3_11_2]
```
I've split the namespace into the current hierarchy in case there's some use for it later but I've not thought of anything yet. That would also simplify the `abi_macros.hpp` code a little.

Do we want an easier option to turn off that inline namespace?
`NLOHMANN_JSON_NAMESPACE_SIMPLE=1` perhaps?
Also, `NLOHMANN_JSON_NAMESPACE_EXCLUDE_VERSION` could be shortened to `NLOHMANN_JSON_NAMESPACE_NO_VERSION`.

Reply:
I'm leaning towards not making turning off the inline namespaces any easier.

Reply:
I think this is probably the clearest as to the reason for the namespace existing, and not duplicating the initial `nlohmann` tag, if we can't come up with a reason to have version as a separate namespace.

    nlohmann::abi_tags[_diag][_ldvcmp][_v3_11_2]

It does however seem weird to end up with just this:

    nlohmann::abi_tags

So maybe `_diag` isn't `""/"_diag"` but `_nodiag/_diag`?  That way there's always at least `nlohmann::abi_tags_nodiag` or `nlohmann::abi_tags_diag` (with similar for `_ldvcmp`?)

Reply:
How about `json_abi` instead of `nlohmann_json`? Keeps the name a bit shorter with no options enabled and is less weird than `abi_tags` alone.

Reply:
Sure.

Reply:
Any objections about shortening `NLOHMANN_JSON_NAMESPACE_EXCLUDE_VERSION` to `NLOHMANN_JSON_NAMESPACE_NO_VERSION`?

Reply:
No objections.

Reply:
Go ahead!